array.go

```diff
@@ -913,3 +913,59 @@ func countElement(array []interface{}, element interface{}) int {
 	}
 	return count
 }
+
+// Transform runs the passed function on all the Elements in the array.
+//
+// If transformation inside function fails, the original Array is marked failed.
+//
+// Transform will immediately stop execution of the function
+// upon first transformation failures
+// or if return value is nil.
+//
+// Example:
+//
+//	array := NewArray(t, []interface{}{"foo", "bar"})
+//	transformedArray := array.Transform(
+//		func(index int, value *httpexpect.Value) *httpexpect.Value {
+//			return *httpexpect.NewValue(...)
+//		},
+//	)
+func (a *Array) Transform(fn func(index int, value *Value) *Value) *Array {
```

Comment:
Given it a second thought, and after looking at real usage example in your tests, it seems that it would be more logical and convenient if transform function operated with raw values, instead of httpexpect.Value.

Why:

* passing httpexpect.Value to user func looks artificial: most times, user will just call `.Raw()` and don't use httpexpect.Value itself, because the point of Transform is to replace the underlying value, not to perform assertions on it

* returning httpexpect.Value, again, looks artificial: user func wants to return underlying value; however it has to wrap it with httpexpect.Value, and then we immediately unwrap it back because we also need only underlying value

So it seems this would a better API:

```go
func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Array
```

(Note that in this case nil return value is not a failure.)

What do you think?

Reply:
I see what you mean. We can unwrap the underlying value and pass it to the function and then wrap the return value and add it to the new Array/Object. However in that case how do the Transform function handle failures? Should we check if the chain's failbit is true and just return? If so then what do we return the copy of original array or the half transformed array?

Reply:
> We can unwrap the underlying value and pass it to the function

Actually the underlying array holds unwrapped values.

> and then wrap the return value and add it to the new Array/Object.

Actually our filteredSlice also holds unwrapped values.

So no need for any wraps / unwraps at all (unlike current implementation).

> However in that case how do the Transform function handle failures?

httpexpect is for testing, so we can assume that transform function is executing inside a test. Thus, it can use regular facilities to report failures in tests - t.Errorf, t.Fatalf, testify.

We don't need to bother I think. Transform function will return something, maybe nil, maybe not. If it will decide to fail the test also, it's not our business, our help is not needed for that.

The reason why we're currently checking for nil return value is just because we can't do anything meaningful if we get nil httpexpect.Value. And returning nil is not the only way to fail, currently transform func can call t.Errorf to mark test failed, and we're not trying to handle it (and don't have to).

Reply:
Hey @gavv I think we don't need 

> If transformation inside function fails, the original Array is marked failed.

This behaviour. We can just let the user handle whether the function failure should be fatal or just a error using t.Fatal/t.Error because in case of t.Error the passed function would still run for all the values. The reason I am saying this is because inside the `Transform` function would not know if the transformation failed.

```go
func (a *Array) Transform(fn func(index int, value interface{}) interface{}) *Array {
	a.chain.enter("Transform()")
	defer a.chain.leave()

	if a.chain.failed() {
		return a
	}

	if fn == nil {
		a.chain.fail(AssertionFailure{
			Type: AssertUsage,
			Errors: []error{
				errors.New("unexpected nil function argument"),
			},
		})
		return newArray(a.chain, nil)
	}

	array := []interface{}{}

	for index, val := range a.value {
		transformedValue := fn(index, val)
		array = append(array, transformedValue)
	}

	return newArray(a.chain, array)
}
```

Notice I removed the chain creation for each value since only raw values will be sent to the function. So the `Transform`
function would not know if there is any error. We can just leave the error handling to the user and let them decide do they want to fail the entire transformation operation or log the error and continue on. 

Does this make sense? What do you think?

Reply:
Absolutely. Looks good

Reply:
Awesome.

Reply:
Hey @gavv I am done with changes. Have a look and lemme know if something else requires improvement.
Cheers.